File: core/python/pubplot/recipes/correlation.py
"""Recipe: correlation between two continuous variables. Mirrors the R recipe:
picks Pearson or Spearman from the normality of each variable (unless forced),
scatter with a linear fit and CI band, coefficient and p annotated."""
import numpy as np
import pandas as pd
import pingouin as pg
import seaborn as sns

from ..bundle import slugify
from ..theme import apply_pub_style, new_fig, palette
from ..util import cap_first, format_p, shapiro_normal


def _resolve(method, x_normal, y_normal):
    method = (method or "auto").lower()
    if method == "auto":
        m = "pearson" if (x_normal and y_normal) else "spearman"
    elif method in ("pearson", "spearman", "kendall"):
        m = method
    else:
        raise ValueError(f"unknown correlation method '{method}' "
                         "(use pearson, spearman, or kendall)")
    label = {"pearson": "Pearson correlation",
             "spearman": "Spearman rank correlation",
             "kendall": "Kendall rank correlation"}[m]
    est_name = {"pearson": "r", "spearman": "rho", "kendall": "tau"}[m]
    return dict(method=m, label=label, estimate_name=est_name)


def recipe_correlation(df, spec):
    x, y = spec["data"]["x"], spec["data"]["y"]
    if not x or not y:
        raise ValueError("correlation needs --x and --y (two numeric columns)")
    for c in (x, y):
        if c not in df.columns:
            raise ValueError(f"column '{c}' not found in data")

    n0 = len(df)
    xv = pd.to_numeric(df[x], errors="coerce")
    yv = pd.to_numeric(df[y], errors="coerce")
    mask = np.isfinite(xv) & np.isfinite(yv)
    xv, yv = xv[mask].to_numpy(), yv[mask].to_numpy()
    df_used = pd.DataFrame({x: xv, y: yv})
    dropped = int((~mask).sum())
    steps = ([f"Dropped {dropped} row(s) with a non-numeric or missing value in "
              f"'{x}' or '{y}' ({n0} -> {len(xv)})."] if dropped
             else ["No cleaning applied; input used as-is."])
    if len(xv) < 3:
        raise ValueError("correlation needs at least 3 complete pairs")

    px, xn = shapiro_normal(xv)
    py, yn = shapiro_normal(yv)
    r = _resolve(spec["test"].get("method"), xn, yn)

    res = pg.corr(xv, yv, method=r["method"])
    estimate = float(res["r"].iloc[0])
    pval = float(res["p_val"].iloc[0])
    if "CI95" in res:
        ci = res["CI95"].iloc[0]
        ci_low, ci_high = float(ci[0]), float(ci[1])
    else:
        ci_low, ci_high = np.nan, np.nan
    if r["method"] == "pearson":
        dof = len(xv) - 2
        statistic = (estimate * np.sqrt(dof / (1 - estimate ** 2))
                     if abs(estimate) < 1 else np.inf)
    else:
        dof, statistic = np.nan, np.nan

    ap = spec["appearance"]
    pal = palette(1, ap.get("palette"))
    w, h = 4.6, 4.2
    fig, ax = new_fig(w, h)
    sns.regplot(x=xv, y=yv, ax=ax, color=pal[0],
                scatter_kws={"s": 24, "alpha": 0.8, "edgecolor": "none"},
                line_kws={"color": "#333333", "linewidth": 1.4})
    pv = format_p(pval)
    est_txt = f"{r['estimate_name']} = {estimate:.2f}, p {'' if pv.startswith('<') else '= '}{pv}"
    ax.annotate(est_txt, xy=(0.03, 0.96), xycoords="axes fraction",
                ha="left", va="top", fontsize=11)
    ax.set_xlabel(ap.get("x_label") or x)
    ax.set_ylabel(ap.get("y_label") or y)
    if ap.get("title"):
        ax.set_title(ap["title"], fontweight="bold")
    yl = ap.get("y_limits") or [None, None]
    if yl[0] is not None or yl[1] is not None:
        ax.set_ylim(yl[0], yl[1])
    apply_pub_style(ax, ap.get("theme"))
    fig.tight_layout()

    stats_df = pd.DataFrame([{
        "x": x, "y": y, "n": len(xv), "method": r["label"],
        "estimate_name": r["estimate_name"], "estimate": estimate,
        "statistic": statistic, "df": dof, "p_value": pval,
        "conf_low": ci_low, "conf_high": ci_high,
        "shapiro_p_x": px, "shapiro_p_y": py,
    }])

    test_meta = {
        "name": r["label"],
        "estimate": {"name": r["estimate_name"], "value": estimate},
        "df": None if np.isnan(dof) else int(dof),
        "p_value": pval,
        "conf_int": {"low": ci_low, "high": ci_high},
        "assumptions": {"normality": {"test": "shapiro-wilk", "p_x": px, "p_y": py}},
    }
    methods = _methods(r, stats_df, spec)
    return {
        "fig": fig, "stats": stats_df, "test_meta": test_meta, "methods": methods,
        "df_used": df_used, "clean_steps": steps,
        "label": slugify(f"{y}_vs_{x}"), "width": w, "height": h,
        "build_script": lambda in_name, fig_stub: _script(spec, r, pal[0], in_name, fig_stub),
    }


def _methods(r, stats_df, spec):
    from importlib.metadata import version as v
    import platform
    x_lab = spec["appearance"].get("x_label") or stats_df["x"].iloc[0]
    y_lab = spec["appearance"].get("y_label") or stats_df["y"].iloc[0]
    auto_txt = (" The method was chosen from the Shapiro-Wilk normality of each variable."
                if (spec["test"].get("method") or "auto").lower() == "auto" else "")
    est = f"{stats_df['estimate'].iloc[0]:.2f}"
    pv = format_p(float(stats_df["p_value"].iloc[0]))
    return (
        f"The association between {cap_first(x_lab)} and {y_lab} was assessed with a "
        f"{r['label']} (n = {int(stats_df['n'].iloc[0])})." + auto_txt +
        f" The correlation was {r['estimate_name']} = {est} "
        f"(P {'' if pv.startswith('<') else '= '}{pv}), with a linear fit and 95% "
        "confidence band shown. Significance was set at P < 0.05."
        f" Analyses were performed in Python {platform.python_version()} with pingouin {v('pingouin')};"
        f" the figure was produced with seaborn {v('seaborn')}."
    )


def _script(spec, r, color, in_name, fig_stub):
    x, y = spec["data"]["x"], spec["data"]["y"]
    ap = spec["appearance"]
    return f'''#!/usr/bin/env python3
# Standalone reproduction. Run inside the pinned pubplot container
# (see REPRODUCE.md) from this bundle folder.
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np, pandas as pd, seaborn as sns, pingouin as pg

raw = pd.read_csv("{in_name}")
x = pd.to_numeric(raw["{x}"], errors="coerce")
y = pd.to_numeric(raw["{y}"], errors="coerce")
mask = np.isfinite(x) & np.isfinite(y)
x, y = x[mask].to_numpy(), y[mask].to_numpy()
print(pg.corr(x, y, method="{r['method']}"))

fig, ax = plt.subplots(figsize=(4.6, 4.2))
sns.regplot(x=x, y=y, ax=ax, color="{color}",
            scatter_kws={{"s": 24, "alpha": 0.8, "edgecolor": "none"}},
            line_kws={{"color": "#333333", "linewidth": 1.4}})
ax.set_xlabel("{ap.get('x_label') or x}"); ax.set_ylabel("{ap.get('y_label') or y}")
for s in ("top", "right"): ax.spines[s].set_visible(False)
fig.tight_layout()
fig.savefig("{fig_stub}.pdf")
print("Reproduced figure.")
'''

File: core/r/R/correlation.R
# Recipe: correlation between two continuous variables.
# Chooses Pearson or Spearman from the normality of each variable (unless the
# caller forces the method), draws a scatter with a linear fit and CI band, and
# annotates the coefficient and p on the plot. Returns the plot, a tidy stats
# table, and test metadata.

# Resolve which correlation method to run from the request and the normality of
# each variable. "auto" uses Pearson only when both variables look normal.
resolve_correlation_method <- function(method, x_normal, y_normal) {
  method <- tolower(method %||% "auto")
  if (method == "auto") {
    m <- if (isTRUE(x_normal) && isTRUE(y_normal)) "pearson" else "spearman"
  } else if (method %in% c("pearson", "spearman", "kendall")) {
    m <- method
  } else {
    stop(sprintf("unknown correlation method '%s' (use pearson, spearman, or kendall)", method))
  }
  label <- switch(m,
    pearson  = "Pearson correlation",
    spearman = "Spearman rank correlation",
    kendall  = "Kendall rank correlation")
  estimate_name <- switch(m, pearson = "r", spearman = "rho", kendall = "tau")
  list(method = m, family = "correlation", label = label, estimate_name = estimate_name)
}

recipe_correlation <- function(df, spec) {
  x <- spec$data$x
  y <- spec$data$y
  if (is.null(x) || is.null(y)) {
    stop("correlation needs --x and --y (two numeric columns)")
  }
  if (!x %in% names(df)) stop(sprintf("column '%s' not found in data", x))
  if (!y %in% names(df)) stop(sprintf("column '%s' not found in data", y))

  # Clean inline: both columns numeric, drop non-finite pairs.
  n0 <- nrow(df)
  xv <- suppressWarnings(as.numeric(df[[x]]))
  yv <- suppressWarnings(as.numeric(df[[y]]))
  keep <- is.finite(xv) & is.finite(yv)
  xv <- xv[keep]; yv <- yv[keep]
  df_used <- data.frame(a = xv, b = yv)
  names(df_used) <- c(x, y)
  clean_steps <- if (sum(!keep) > 0) {
    sprintf("Dropped %d row(s) with a non-numeric or missing value in '%s' or '%s' (%d -> %d).",
            sum(!keep), x, y, n0, length(xv))
  } else "No cleaning applied; input used as-is."
  if (length(xv) < 3) stop("correlation needs at least 3 complete pairs")

  x_norm <- shapiro_safe(xv)
  y_norm <- shapiro_safe(yv)
  resolved <- resolve_correlation_method(spec$test$method, x_norm$normal, y_norm$normal)

  ct <- stats::cor.test(xv, yv, method = resolved$method)
  estimate  <- unname(ct$estimate)
  ci        <- if (!is.null(ct$conf.int)) as.numeric(ct$conf.int) else c(NA_real_, NA_real_)
  statistic <- if (!is.null(ct$statistic)) unname(ct$statistic) else NA_real_
  dfree     <- if (!is.null(ct$parameter)) unname(ct$parameter) else NA_real_

  pal <- spec$appearance$palette %||% pubplot_palette
  point_col <- pal[1]

  p <- ggplot2::ggplot(df_used, ggplot2::aes(x = .data[[x]], y = .data[[y]])) +
    ggplot2::geom_point(colour = point_col, size = 2, alpha = 0.8) +
    ggplot2::geom_smooth(method = "lm", formula = y ~ x, se = TRUE,
                         colour = "#333333", fill = "grey75", linewidth = 0.6) +
    ggpubr::stat_cor(method = resolved$method, label.x.npc = "left",
                     label.y.npc = "top", size = 3.6) +
    ggplot2::labs(
      x     = spec$appearance$x_label %||% x,
      y     = spec$appearance$y_label %||% y,
      title = spec$appearance$title) +
    pub_theme(spec$appearance$theme)

  ylim <- spec$appearance$y_limits
  ymin <- if (length(ylim) >= 1) ylim[[1]] else NULL
  ymax <- if (length(ylim) >= 2) ylim[[2]] else NULL
  if (!is.null(ymin) || !is.null(ymax)) {
    p <- p + ggplot2::coord_cartesian(ylim = c(ymin, ymax))
  }

  stats_df <- data.frame(
    x             = x,
    y             = y,
    n             = length(xv),
    method        = resolved$label,
    estimate_name = resolved$estimate_name,
    estimate      = estimate,
    statistic     = statistic,
    df            = dfree,
    p_value       = unname(ct$p.value),
    conf_low      = ci[1],
    conf_high     = ci[2],
    shapiro_p_x   = x_norm$p,
    shapiro_p_y   = y_norm$p,
    stringsAsFactors = FALSE)

  test_meta <- list(
    name      = resolved$label,
    estimate  = list(name = resolved$estimate_name, value = estimate),
    statistic = statistic,
    df        = dfree,
    p_value   = unname(ct$p.value),
    conf_int  = list(low = ci[1], high = ci[2]),
    assumptions = list(
      normality = list(test = "shapiro-wilk", p_x = x_norm$p, p_y = y_norm$p)))

  methods <- .cor_methods(resolved, stats_df, spec)
  build_script <- function(in_name, fig_stub) .cor_emit_script(spec, resolved, point_col, in_name, fig_stub)

  list(plot = p, stats = stats_df, test_meta = test_meta, resolved = resolved,
       methods = methods, build_script = build_script,
       label = slugify(sprintf("%s_vs_%s", y, x)),
       df_used = df_used, clean_steps = clean_steps,
       qc = list(quantity = "variable values",
                 panels = stats::setNames(list(xv, yv), c(x, y))),
       width = 4.6, height = 4.2)
}

.cor_methods <- function(resolved, stats_df, spec) {
  v    <- pkg_versions(c("ggplot2", "ggpubr"))
  rver <- paste(R.version$major, R.version$minor, sep = ".")
  x_lab <- spec$appearance$x_label %||% stats_df$x[1]
  y_lab <- spec$appearance$y_label %||% stats_df$y[1]
  auto_txt <- if (tolower(spec$test$method %||% "auto") == "auto") {
    " The method was chosen from the Shapiro-Wilk normality of each variable."
  } else ""
  est <- formatC(stats_df$estimate[1], format = "f", digits = 2)
  pv  <- formatC(stats_df$p_value[1], format = "g", digits = 2)
  paste0(
    sprintf("The association between %s and %s was assessed with a %s (n = %d).",
            cap_first(x_lab), y_lab, resolved$label, stats_df$n[1]),
    auto_txt,
    sprintf(" The correlation was %s = %s (P = %s), with a linear fit and 95%% confidence band shown.",
            resolved$estimate_name, est, pv),
    " Significance was set at P < 0.05.",
    sprintf(" Rendered in R %s with ggplot2 %s and ggpubr %s.",
            rver, v$ggplot2, v$ggpubr))
}

.cor_emit_script <- function(spec, resolved, point_col, in_name, fig_stub) {
  x <- spec$data$x; y <- spec$data$y
  xlab <- spec$appearance$x_label %||% x
  ylab <- spec$appearance$y_label %||% y
  c(
    "#!/usr/bin/env Rscript",
    "# Standalone reproduction. Run inside the pinned pubplot container",
    "# (see REPRODUCE.md) from this bundle folder.",
    "suppressPackageStartupMessages({ library(ggplot2); library(ggpubr) })",
    "",
    sprintf('raw <- read.csv("%s", check.names = FALSE)', in_name),
    sprintf('x <- suppressWarnings(as.numeric(raw[["%s"]]))', x),
    sprintf('y <- suppressWarnings(as.numeric(raw[["%s"]]))', y),
    "keep <- is.finite(x) & is.finite(y)",
    "df <- data.frame(x = x[keep], y = y[keep])",
    sprintf('ct <- cor.test(df$x, df$y, method = "%s")', resolved$method),
    "print(ct)",
    "",
    "p <- ggplot(df, aes(x = x, y = y)) +",
    sprintf('  geom_point(colour = "%s", size = 2, alpha = 0.8) +', point_col),
    '  geom_smooth(method = "lm", formula = y ~ x, se = TRUE, colour = "#333333", fill = "grey75") +',
    sprintf('  stat_cor(method = "%s", label.x.npc = "left", label.y.npc = "top") +', resolved$method),
    sprintf('  labs(x = "%s", y = "%s") + theme_classic()', xlab, ylab),
    sprintf('ggsave("%s.pdf", p, width = 4.6, height = 4.2)', fig_stub),
    'cat("Reproduced figure.\\n")'
  )
}
